### PR TITLE
Horovod multi-node multi-GPU support in pre-training

### PR DESCRIPTION
**Pull request by c0redumb**
This code added Horovod multi-node multi-GPU support in pre-training. The Horovod feature can be turned on/off by setting "--use_horovod" flag (default to False).

### CODE REVIEW

**Comment by c0redumb** [comment]
I think I will also need to update the README file to describe the changes we have made and how to use them.

File: README.md [hunk 1]
```diff
@@ -1,5 +1,31 @@
...
+To run this script, do
+```shellscript
+scripts/run_model_in_rocm_docker.sh
+```
+### B. Horovod support
+
+Horovod scaling is enabled in the pre-training code. To use the Horovod scaling,
+please run `pre_training.py` with the `--use_horovod` commandline options set to `True`.
+If left unset, the pre-training defaults to the original code without Horovod.
+
```
[the remark this answers is not in the document]
Comment (micmelesse) [reply README.md]: If not then add info on how to call the scripts? also in what docker container?

File: optimization.py [hunk 2]
```diff
@@ -36,6 +40,9 @@ def create_optimizer(loss, init_lr, num_train_steps, num_warmup_steps, use_tpu):
       end_learning_rate=0.0,
       power=1.0,
       cycle=False)
+  # if use_hvd:
+  #   # May want to scale learning rate by number of GPUs
+  #   learning_rate *= hvd.size()
```
[the remark this answers is not in the document]
Comment (c0redumb) [reply optimization.py]: Lost in the shuffle. I will put it back.

File: optimization.py [hunk 3]
```diff
@@ -63,18 +70,27 @@ def create_optimizer(loss, init_lr, num_train_steps, num_warmup_steps, use_tpu):
       beta_2=0.999,
       epsilon=1e-6,
       exclude_from_weight_decay=["LayerNorm", "layer_norm", "bias"])
+  
+  if use_hvd:
```
[the remark this answers is not in the document]
Comment (c0redumb) [reply optimization.py]: I think they are mutually exclusive, but not either/or relationship. For example, it can be non-TPU and non-Horovod case where there is a single GPU used.

File: optimization.py [hunk 4]
```diff
@@ -63,18 +70,27 @@ def create_optimizer(loss, init_lr, num_train_steps, num_warmup_steps, use_tpu):
...
+    # [HVD] Use distributed optimizer to compute gradients
+    grads_and_vars=optimizer.compute_gradients(loss, tvars)
+    grads = [grad for grad,var in grads_and_vars]
+    tvars = [var for grad,var in grads_and_vars]
+  else:
+    # Use standard TF gradients
+    grads = tf.gradients(loss, tvars)
 
-  # This is how the model was pre-trained.
```
[the remark this answers is not in the document]
Comment (c0redumb) [reply optimization.py]: Lost in shuffle. I will put it back.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (c0redumb) [reply README.md]: You are right. The issue is that this script needs to be invoked in a Horovod OpenMPI environment. At this time, it is done on a private container. Let's think about how to present that.

File: scripts/run_model_in_rocm_hvd_docker.sh [hunk 5]
```diff
@@ -0,0 +1,135 @@
+#!/bin/bash
+
+MODEL="BERT-Large"
+#IMAGE="rocm/tensorflow:latest"
+IMAGE="rocm/tensorflow:rocm3.1-tf1.15-dev"
```
Comment (micmelesse) [inline scripts/run_model_in_rocm_hvd_docker.sh]: does this image have horovad?

Comment (micmelesse) [reply scripts/run_model_in_rocm_hvd_docker.sh]: I read your comment in the readme.

Comment (c0redumb) [reply scripts/run_model_in_rocm_hvd_docker.sh]: Unfortunately, there is no public available container at this time. So this script is provided as a future reference here.

**Comment by c0redumb** [comment]
I did a million step run. The result looked strange. The loss has only gone down to the 7. I will start a few more runs to see if there is something wrong with it.
